Lib/multiprocessing/context.py

```diff
@@ -200,15 +227,34 @@ def get_start_method(self, allow_none=False):
     def set_start_method(self, method, force=False):
         raise ValueError('cannot set start method of concrete context')
 
-    @property
-    def reducer(self):
+    def get_reducer(self):
         '''Controls how objects will be reduced to a form that can be
         shared with other processes.'''
-        return globals().get('reduction')
+        reducer = self._reducer
+        if reducer is None:
+            return reduction
+        return reducer
+
+    def set_reducer(self, reduction):
+        if reduction is original_reducer:
+            self._reducer = None
+            return
 
-    @reducer.setter
-    def reducer(self, reduction):
-        globals()['reduction'] = reduction
+        if not isinstance(reduction, original_reducer.AbstractReducer):
```

Comment:
The checks below aren't strict enough; they pass for a `.register()`ed subclass of `Abstract<Reducer|Pickler|Unpickler>`, but the code won't actually work with such a class due to depending on undocumented APIs.

Reply:
I do not understand what you mean by "Undocumented API". This PR comes with documentation for the `AbstractReducer`, `AbstractPickler`, and `AbstractUnpickler`.

Reply:
What I mean is that the documentation suggests that the below code should work:
``` 
class DummyReducer:
    def get_pickler_class():
        return pickle.Pickler
    def get_unpickler_class():
        return pickle.Unpickler
AbstractReducer.register(DummyReducer)
multiprocessing.set_reducer(DummyReducer())
```
However, in practice, it appears that running that code (and then doing something that uses reduction) will produce a (rather cryptic) error message in some cases. Either the documentation should be clear that virtual subclasses don't work and the code should raise an error when a virtual subclass is passed, or the code should be updated to handle virtual subclasses.

Reply:
> What I mean is that the documentation suggests that the below code should work

Not really: the `get_pickler_class` API documentation specifies that `get_pickler_class` must return an`AbstractPickler` subclass, but your `DummyReducer` simply returns a `Pickler`, which is not an `AbstractPickler` subclass.

Reply:
You didn't misunderstand anything, my example was accidentally bogus, and I misunderstood the code.  The actual bug I was trying to point out appears not to exist with reducers, but instead with picklers.
```python
class DummyReducer(AbstractReducer):
    def get_pickler_class():
        return DummyPickler
class DummyPickler(pickle.Pickler):
    pass
AbstractPickler.register(pickle.Pickler)
multiprocessing.set_reducer(DummyReducer()
```
This seemingly harmless code has the effect of breaking the passing of all of the custom types that `multiprocessing` registers, in an entirely non-intuitive way, which I think should either be fixed or documented.